src/exceptions.rs

```diff
@@ -613,24 +615,62 @@ impl_windows_native_exception!(
 );
 
 impl PyUnicodeDecodeError {
-    /// Creates a Python `UnicodeDecodeError`.
+    /// Deprecated form of [`PyUnicodeDecodeError::new_bound`].
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`PyUnicodeDecodeError::new` will be replaced by `PyUnicodeDecodeError::new_bound` in a future PyO3 version"
+        )
+    )]
     pub fn new<'p>(
         py: Python<'p>,
         encoding: &CStr,
         input: &[u8],
         range: ops::Range<usize>,
         reason: &CStr,
     ) -> PyResult<&'p PyUnicodeDecodeError> {
-        unsafe {
-            py.from_owned_ptr_or_err(ffi::PyUnicodeDecodeError_Create(
-                encoding.as_ptr(),
-                input.as_ptr() as *const c_char,
-                input.len() as ffi::Py_ssize_t,
-                range.start as ffi::Py_ssize_t,
-                range.end as ffi::Py_ssize_t,
-                reason.as_ptr(),
-            ))
-        }
+        Ok(PyUnicodeDecodeError::new_bound(py, encoding, input, range, reason)?.into_gil_ref())
+    }
+
+    /// Creates a Python `UnicodeDecodeError`.
+    pub fn new_bound<'p>(
+        py: Python<'p>,
+        encoding: &CStr,
+        input: &[u8],
+        range: ops::Range<usize>,
+        reason: &CStr,
+    ) -> PyResult<Bound<'p, PyUnicodeDecodeError>> {
+        let bound = unsafe {
+            Bound::from_owned_ptr_or_err(
+                py,
+                ffi::PyUnicodeDecodeError_Create(
+                    encoding.as_ptr(),
+                    input.as_ptr() as *const c_char,
+                    input.len() as ffi::Py_ssize_t,
+                    range.start as ffi::Py_ssize_t,
+                    range.end as ffi::Py_ssize_t,
+                    reason.as_ptr(),
+                ),
+            )
+        };
+        Ok(bound?.downcast_into()?)
```

Comment:
Similar golfing here, using some internal helper traits:

```suggestion
        unsafe {
            ffi::PyUnicodeDecodeError_Create(
                encoding.as_ptr(),
                input.as_ptr() as *const c_char,
                input.len() as ffi::Py_ssize_t,
                range.start as ffi::Py_ssize_t,
                range.end as ffi::Py_ssize_t,
                reason.as_ptr(),
            )
            .assume_owned_or_err(py)
        }.downcast_into()
```